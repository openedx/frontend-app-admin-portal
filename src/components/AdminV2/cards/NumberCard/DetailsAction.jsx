import { Link, useLocation, useParams } from 'react-router-dom';
import { Icon, Spinner } from '@openedx/paragon';
import classNames from 'classnames';
import { ArrowDownward } from '@openedx/paragon/icons';
import PropTypes from 'prop-types';
import React from 'react';
import { removeTrailingSlash } from '../../../../utils';

const DetailsAction = ({
  slug, label, isLoading, onClick,
}) => {
  const location = useLocation();
  const { actionSlug } = useParams();
  const slugIndex = location?.pathname.lastIndexOf('/');
  const truncatedURL = location?.pathname.substring(0, slugIndex);

  return (
    <Link
      key={label}
      to={actionSlug ? `${truncatedURL}/${slug}` : `${removeTrailingSlash(location?.pathname)}/${slug}`}
      onClick={onClick}
    >
      <div
        className={classNames('p-2 d-flex align-items-center justify-content-between', { 'bg-primary-100': slug === actionSlug })}
      >
        <span className="label">{label}</span>
        <Icon src={ArrowDownward} />
        {isLoading && <Spinner animation="border" className="ml-2" variant="primary" size="sm" />}
      </div>
    </Link>
  );
};

DetailsAction.propTypes = {
  slug: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  isLoading: PropTypes.bool.isRequired,
  onClick: PropTypes.func.isRequired,
};

export default DetailsAction;
