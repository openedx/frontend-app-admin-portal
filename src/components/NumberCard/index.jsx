import React from 'react';
import classNames from 'classnames';
import PropTypes from 'prop-types';
import { Button, Icon, Spinner } from '@edx/paragon';
import { ArrowDropDown, Close } from '@edx/paragon/icons';
import { Link, withRouter } from 'react-router-dom';

import { removeTrailingSlash, isTriggerKey } from '../../utils';
import { DETAILS_TEXT } from '../CouponDetails/constants';

export const triggerKeys = {
  OPEN_DETAILS: ['ArrowDown', 'Enter'],
  CLOSE_DETAILS: ['Escape'],
  NAVIGATE_DOWN: ['ArrowDown', 'Tab'],
  NAVIGATE_UP: ['ArrowUp'],
  SELECT_ACTION: ['Enter', ' '],
};

class NumberCard extends React.Component {
  constructor(props) {
    super(props);

    this.detailActionItemRefs = [];
    this.toggleDetailsBtnRef = null;
    this.containerRef = null;

    this.state = {
      detailsExpanded: false,
      focusIndex: 0,
    };

    this.toggleDetails = this.toggleDetails.bind(this);
    this.handleDetailsActionClick = this.handleDetailsActionClick.bind(this);
    this.handleToggleDetailsKeyDown = this.handleToggleDetailsKeyDown.bind(this);
    this.handleDocumentClick = this.handleDocumentClick.bind(this);
  }

  componentDidUpdate(prevProps, prevState) {
    const { focusIndex } = this.state;
    const { detailsExpanded } = this.props;

    if (detailsExpanded !== prevProps.detailsExpanded) {
      this.setState({ // eslint-disable-line react/no-did-update-set-state
        detailsExpanded,
      });
    }

    if (focusIndex !== prevState.focusIndex) {
      this.detailActionItemRefs[this.state.focusIndex].focus();
    }
  }

  handleDocumentClick(e) {
    if (this.containerRef && this.containerRef.contains(e.target)
      && this.containerRef !== e.target) {
      return;
    }
    this.toggleDetails();
  }

  handleDetailsActionClick(event) {
    if (event) {
      event.target.click();
    }
    this.toggleDetails();
  }

  handleToggleDetailsKeyDown(event) {
    const { detailsExpanded } = this.state;
    if (!detailsExpanded && isTriggerKey({ triggerKeys, action: 'OPEN_DETAILS', key: event.key })) {
      event.preventDefault();
      this.toggleDetails();
    } else if (detailsExpanded && isTriggerKey({ triggerKeys, action: 'CLOSE_DETAILS', key: event.key })) {
      this.toggleDetails();
    }
  }

  handleDetailsActionKeyDown(event) {
    const { focusIndex } = this.state;
    const { detailActions } = this.props;

    event.preventDefault();

    if (isTriggerKey({ triggerKeys, action: 'CLOSE_DETAILS', key: event.key })) {
      this.toggleDetails();
    } else if (isTriggerKey({ triggerKeys, action: 'SELECT_ACTION', key: event.key })) {
      this.handleDetailsActionClick(event);
    } else if (isTriggerKey({ triggerKeys, action: 'NAVIGATE_DOWN', key: event.key })) {
      this.setState({
        focusIndex: (focusIndex + 1) % detailActions.length,
      });
    } else if (isTriggerKey({ triggerKeys, action: 'NAVIGATE_UP', key: event.key })) {
      this.setState({
        focusIndex: ((focusIndex - 1) + detailActions.length) % detailActions.length,
      });
    }
  }

  toggleDetails() {
    this.setState((state) => ({
      detailsExpanded: !state.detailsExpanded,
      focusIndex: 0,
    }), () => {
      // Wait until after the state is set for the DOM elements
      // to render before we set focus.
      if (this.state.detailsExpanded) {
        this.detailActionItemRefs[this.state.focusIndex].focus();
        this.addEvents();
      } else {
        this.toggleDetailsBtnRef.focus();
        this.removeEvents();
      }
    });
  }

  formatTitle(title) {
    if (typeof title === 'number') {
      return title.toLocaleString();
    }
    return title;
  }

  addEvents() {
    document.addEventListener('click', this.handleDocumentClick, true);
  }

  removeEvents() {
    document.removeEventListener('click', this.handleDocumentClick, true);
  }

  renderDetailActions() {
    const { detailActions, match } = this.props;
    const { params: { actionSlug } } = match;

    return detailActions.map((action, index) => (
      <Link
        innerRef={(node) => { this.detailActionItemRefs[index] = node; }}
        className={classNames(
          'btn btn-link',
          {
            active: action.slug === actionSlug,
          },
        )}
        key={action.label}
        to={actionSlug ? action.slug : `${removeTrailingSlash(match.url)}/${action.slug}`}
        onClick={() => { this.handleDetailsActionClick(); }}
        onKeyDown={event => this.handleDetailsActionKeyDown(event)}
      >
        <div className="d-flex justify-content-between align-items-center">
          <span className="label">
            {action.label}
          </span>
          {action.loading
            && <Spinner animation="border" className="ml-2" variant="primary" size="sm" />}
        </div>
      </Link>
    ));
  }

  render() {
    const { detailsExpanded } = this.state;
    const {
      className,
      title,
      icon,
      description,
      detailActions,
      id,
    } = this.props;

    return (
      <div
        ref={(node) => { this.containerRef = node; }}
        className={classNames('number-card w-100', {
          'has-details': detailActions,
        })}
      >
        <div className={classNames('card', className)}>
          <div className="card-body">
            <h3 className="card-title d-flex align-items-center justify-content-between">
              <span>
                {this.formatTitle(title)}
              </span>
              {icon && (
                <Icon
                  className={
                    classNames(
                      'd-flex align-items-center justify-content-center',
                    )
                  }
                  src={icon}
                />
              )}
            </h3>
            <p className="card-text">{description}</p>
          </div>
        </div>
        {detailActions && (
          <div className="card-footer">
            <div className="footer-title">
              <Button
                variant="link"
                ref={(node) => { this.toggleDetailsBtnRef = node; }}
                className="toggle-collapse btn-block"
                onClick={this.toggleDetails}
                onKeyDown={this.handleToggleDetailsKeyDown}
                aria-expanded={detailsExpanded}
                aria-controls={`footer-body-${id}`}
              >
                <div className="d-flex justify-content-between align-items-center">
                  <div className="details-btn-text mr-1">
                    {detailsExpanded ? DETAILS_TEXT.expanded : DETAILS_TEXT.unexpanded}
                  </div>
                  <div>
                    <Icon
                      src={!detailsExpanded ? ArrowDropDown : Close}
                      screenReaderText={detailsExpanded
                        ? DETAILS_TEXT.expandedScreenReader : DETAILS_TEXT.unexpandedScreenReader}
                    />
                  </div>
                </div>
              </Button>
            </div>
            <div
              id={`footer-body-${id}`}
              className={classNames(
                'footer-body mt-1',
                {
                  'd-none': !detailsExpanded,
                },
              )}
            >
              {this.renderDetailActions()}
            </div>
          </div>
        )}
      </div>
    );
  }
}

NumberCard.defaultProps = {
  className: null,
  icon: null,
  detailActions: null,
  detailsExpanded: false,
};

NumberCard.propTypes = {
  title: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  description: PropTypes.string.isRequired,
  id: PropTypes.string.isRequired,
  className: PropTypes.string,
  icon: PropTypes.func,
  detailActions: PropTypes.arrayOf(PropTypes.shape({
    label: PropTypes.string,
    loading: PropTypes.bool,
  })),
  detailsExpanded: PropTypes.bool,
  match: PropTypes.shape({
    url: PropTypes.string,
    params: PropTypes.shape({
      actionSlug: PropTypes.string,
    }),
  }).isRequired,
};

export default withRouter(NumberCard);
