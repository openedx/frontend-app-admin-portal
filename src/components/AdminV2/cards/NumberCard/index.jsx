import React from 'react';
import PropTypes from 'prop-types';
import {
  Icon, Card, Collapsible,
} from '@openedx/paragon';

import { FormattedMessage } from '@edx/frontend-platform/i18n';
import DetailsAction from './DetailsAction';

const NumberCard = ({
  title, icon, description, detailActions,
}) => {
  const handleDetailsActionClick = () => {
    const element = document.getElementById('learner-progress-report');
    if (element) {
      requestAnimationFrame(() => {
        element.scrollIntoView({ behavior: 'smooth' });
      });
    }
  };

  return (
    <Card>
      <Card.Header
        title={(
          <div className="d-flex align-items-center justify-content-between">
            <h2 className="h2">
              {typeof title === 'number' ? title.toLocaleString() : title}
            </h2>
            {icon && <Icon src={icon} />}
          </div>
        )}
        subtitle={description}
      />
      <Card.Section>
        <Collapsible
          styling="basic"
          title={(
            <FormattedMessage
              id="adminPortal.cards.collapsible.details"
              defaultMessage="Details"
            />
          )}
        >
          {
            detailActions?.map(action => (
              <DetailsAction
                slug={action.slug}
                label={action.label}
                isLoading={action.isLoading}
                key={action.label}
                onClick={handleDetailsActionClick}
              />
            ))
          }
        </Collapsible>
      </Card.Section>
    </Card>
  );
};

NumberCard.defaultProps = {
  icon: null,
  detailActions: null,
};

NumberCard.propTypes = {
  title: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  description: PropTypes.string.isRequired,
  icon: PropTypes.func,
  detailActions: PropTypes.arrayOf(PropTypes.shape({
    slug: PropTypes.string,
    label: PropTypes.string,
    isLoading: PropTypes.bool,
  })),
};

export default NumberCard;
