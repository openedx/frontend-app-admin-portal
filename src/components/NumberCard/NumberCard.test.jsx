import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';

import NumberCard from './index';

const getNumberCard = wrapper => wrapper.find('NumberCard');

const NumberCardWrapper = props => (
  <MemoryRouter>
    <NumberCard
      id="test-id"
      className="test-class"
      title={10}
      description="This describes the data!"
      iconClassName="fa fa-users"
      {...props}
    />
  </MemoryRouter>
);

describe('<NumberCard />', () => {
  let wrapper;

  describe('renders correctly', () => {
    it('without detail actions', () => {
      const tree = renderer
        .create((
          <NumberCardWrapper />
        ))
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('with detail actions', () => {
      const tree = renderer
        .create((
          <NumberCardWrapper
            detailActions={[{
              label: 'Action 1',
              slug: 'action-1',
            }, {
              label: 'Action 2',
              slug: 'action-2',
            }]}
          />
        ))
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  it('sets a non-number title', () => {
    wrapper = mount(<NumberCardWrapper title="10%" />).find('NumberCard');
    expect(wrapper.find('.card-title span').first().text()).toEqual('10%');
  });

  describe('collapsible detail actions', () => {
    beforeEach(() => {
      wrapper = mount((
        <NumberCardWrapper
          detailActions={[{
            label: 'Action 1',
            slug: 'action-1',
          }, {
            label: 'Action 2',
            slug: 'action-2',
          }]}
        />
      ));
    });

    it('expands and collapses the actions', () => {
      expect(getNumberCard(wrapper).find('.details-btn-text').text()).toEqual('Details');

      wrapper.setProps({ detailsExpanded: true });
      expect(getNumberCard(wrapper).find('.details-btn-text').text()).toEqual('Detailed breakdown');
      expect(getNumberCard(wrapper).instance().state.detailsExpanded).toBeTruthy();

      wrapper.setProps({ detailsExpanded: false });
      expect(getNumberCard(wrapper).find('.details-btn-text').text()).toEqual('Details');
      expect(getNumberCard(wrapper).instance().state.detailsExpanded).toBeFalsy();
    });

    it('opens and closes detail actions with keydown on collapse btn', () => {
      getNumberCard(wrapper).find('.toggle-collapse').hostNodes().simulate('keyDown', { key: 'ArrowDown' });
      expect(getNumberCard(wrapper).instance().state.detailsExpanded).toBeTruthy();

      getNumberCard(wrapper).find('.toggle-collapse').hostNodes().simulate('keyDown', { key: 'Escape' });
      expect(getNumberCard(wrapper).instance().state.detailsExpanded).toBeFalsy();
    });

    it('closes detail actions with enter keydown on action', () => {
      wrapper.setProps({ detailsExpanded: true });
      expect(getNumberCard(wrapper).instance().state.detailsExpanded).toBeTruthy();
      const actions = getNumberCard(wrapper).find('.footer-body .btn-link').hostNodes();
      actions.first().simulate('keyDown', { key: 'Enter' });
      expect(getNumberCard(wrapper).instance().state.detailsExpanded).toBeFalsy();
    });

    it('closes detail actions with escape keydown on action', () => {
      wrapper.setProps({ detailsExpanded: true });
      expect(getNumberCard(wrapper).instance().state.detailsExpanded).toBeTruthy();
      const actions = getNumberCard(wrapper).find('.footer-body .btn-link').hostNodes();
      actions.first().simulate('keyDown', { key: 'Escape' });
      expect(getNumberCard(wrapper).instance().state.detailsExpanded).toBeFalsy();
    });

    it('sets focus index on actions', () => {
      wrapper.setProps({ detailsExpanded: true });
      const actions = getNumberCard(wrapper).find('.footer-body .btn-link').hostNodes();
      actions.first().simulate('keyDown', { key: 'ArrowDown' });
      expect(getNumberCard(wrapper).instance().state.focusIndex).toEqual(1);

      actions.last().simulate('keyDown', { key: 'ArrowUp' });
      expect(getNumberCard(wrapper).instance().state.focusIndex).toEqual(0);
    });

    it('closes detail actions on action click', () => {
      wrapper.setProps({ detailsExpanded: true });
      const actions = getNumberCard(wrapper).find('.footer-body .btn-link').hostNodes();
      actions.first().simulate('click');
      expect(getNumberCard(wrapper).instance().state.detailsExpanded).toBeFalsy();
    });

    it('shows loading spinner in action', () => {
      wrapper = mount((
        <NumberCardWrapper
          detailActions={[{
            label: 'Action 1',
            slug: 'action-1',
            loading: true,
          }]}
        />
      ));

      wrapper.setProps({ detailsExpanded: true });
      const action = getNumberCard(wrapper).find('.footer-body .btn-link').hostNodes().first();
      expect(action.find('.fa-spinner').exists()).toBeTruthy();
    });
  });
});
