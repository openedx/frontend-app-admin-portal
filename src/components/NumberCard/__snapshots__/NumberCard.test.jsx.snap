// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NumberCard /> renders correctly with detail actions 1`] = `
<div
  className="number-card w-100 has-details"
>
  <div
    className="card test-class"
  >
    <div
      className="card-body"
    >
      <h3
        className="card-title d-flex align-items-center justify-content-between"
      >
        <span>
          10
        </span>
        <span
          className="pgn__icon d-flex align-items-center justify-content-center"
        >
          <svg
            aria-hidden={true}
            fill="none"
            focusable={false}
            height={24}
            role="img"
            viewBox="0 0 24 24"
            width={24}
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12 12.75c1.63 0 3.07.39 4.24.9 1.08.48 1.76 1.56 1.76 2.73V18H6v-1.61c0-1.18.68-2.26 1.76-2.73 1.17-.52 2.61-.91 4.24-.91zM4 13c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm1.13 1.1c-.37-.06-.74-.1-1.13-.1-.99 0-1.93.21-2.78.58A2.01 2.01 0 000 16.43V18h4.5v-1.61c0-.83.23-1.61.63-2.29zM20 13c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm4 3.43c0-.81-.48-1.53-1.22-1.85A6.95 6.95 0 0020 14c-.39 0-.76.04-1.13.1.4.68.63 1.46.63 2.29V18H24v-1.57zM12 6c1.66 0 3 1.34 3 3s-1.34 3-3 3-3-1.34-3-3 1.34-3 3-3z"
              fill="currentColor"
            />
          </svg>
        </span>
      </h3>
      <p
        className="card-text"
      >
        This describes the data!
      </p>
    </div>
  </div>
  <div
    className="card-footer"
  >
    <div
      className="footer-title"
    >
      <button
        aria-controls="footer-body-test-id"
        aria-expanded={false}
        className="toggle-collapse btn-block btn btn-link"
        disabled={false}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="button"
      >
        <div
          className="d-flex justify-content-between align-items-center"
        >
          <div
            className="details-btn-text mr-1"
          >
            Details
          </div>
          <div>
            <span
              className="pgn__icon"
            >
              <svg
                aria-hidden={true}
                fill="none"
                focusable={false}
                height={24}
                role="img"
                viewBox="0 0 24 24"
                width={24}
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M7 10l5 5 5-5H7z"
                  fill="currentColor"
                />
              </svg>
              <span
                className="sr-only"
              >
                Show details
              </span>
            </span>
          </div>
        </div>
      </button>
    </div>
    <div
      className="footer-body mt-1 d-none"
      id="footer-body-test-id"
    >
      <a
        className="btn btn-link"
        href="/action-1"
        onClick={[Function]}
        onKeyDown={[Function]}
      >
        <div
          className="d-flex justify-content-between align-items-center"
        >
          <span
            className="label"
          >
            Action 1
          </span>
        </div>
      </a>
      <a
        className="btn btn-link"
        href="/action-2"
        onClick={[Function]}
        onKeyDown={[Function]}
      >
        <div
          className="d-flex justify-content-between align-items-center"
        >
          <span
            className="label"
          >
            Action 2
          </span>
        </div>
      </a>
    </div>
  </div>
</div>
`;

exports[`<NumberCard /> renders correctly without detail actions 1`] = `
<div
  className="number-card w-100"
>
  <div
    className="card test-class"
  >
    <div
      className="card-body"
    >
      <h3
        className="card-title d-flex align-items-center justify-content-between"
      >
        <span>
          10
        </span>
        <span
          className="pgn__icon d-flex align-items-center justify-content-center"
        >
          <svg
            aria-hidden={true}
            fill="none"
            focusable={false}
            height={24}
            role="img"
            viewBox="0 0 24 24"
            width={24}
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12 12.75c1.63 0 3.07.39 4.24.9 1.08.48 1.76 1.56 1.76 2.73V18H6v-1.61c0-1.18.68-2.26 1.76-2.73 1.17-.52 2.61-.91 4.24-.91zM4 13c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm1.13 1.1c-.37-.06-.74-.1-1.13-.1-.99 0-1.93.21-2.78.58A2.01 2.01 0 000 16.43V18h4.5v-1.61c0-.83.23-1.61.63-2.29zM20 13c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm4 3.43c0-.81-.48-1.53-1.22-1.85A6.95 6.95 0 0020 14c-.39 0-.76.04-1.13.1.4.68.63 1.46.63 2.29V18H24v-1.57zM12 6c1.66 0 3 1.34 3 3s-1.34 3-3 3-3-1.34-3-3 1.34-3 3-3z"
              fill="currentColor"
            />
          </svg>
        </span>
      </h3>
      <p
        className="card-text"
      >
        This describes the data!
      </p>
    </div>
  </div>
</div>
`;
